osquery/logger/plugins/tests/aws_logger_tests.cpp

```diff
@@ -0,0 +1,168 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#include <chrono>
+#include <iostream>
+#include <memory>
+#include <vector>
+
+#include <gtest/gtest.h>
+
+#include <osquery/core.h>
+#include <osquery/dispatcher.h>
+#include <osquery/logger.h>
+
+#include "osquery/logger/plugins/aws_log_forwarder.h"
+#include "osquery/tests/test_util.h"
+
+using namespace testing;
+
+namespace osquery {
+class AwsLoggerTests : public testing::Test {
+ public:
+  void SetUp() override {}
+};
+
+class DummyLogForwarder final : public IKinesisLogForwarder {
+ public:
+  DummyLogForwarder() : IKinesisLogForwarder("dummy", 1, 50) {}
+
+  using RawBatch = std::vector<std::string>;
+  using RawBatchList = std::vector<RawBatch>;
+
+  RawBatchList emitted_batch_list_;
+
+ protected:
+  Status internalSetup() override {
+    return Status(0, "OK");
+  }
+
+  Outcome internalSend(const Batch& batch) override {
+    RawBatch raw_batch;
+
+    for (const auto& record : batch) {
+      std::string buffer(
+          reinterpret_cast<const char*>(record.GetData().GetUnderlyingData()),
+          record.GetData().GetLength());
+      raw_batch.push_back(buffer);
+    }
+
+    emitted_batch_list_.push_back(raw_batch);
+
+    return Outcome();
+  }
+
+  void initializeRecord(Record& record,
+                        Aws::Utils::ByteBuffer& buffer) const override {
+    record.SetData(buffer);
+  }
+
+  std::size_t getMaxBytesPerRecord() const override {
+    return 80U;
+  }
+
+  std::size_t getMaxRecordsPerBatch() const override {
+    return 3U;
+  }
+
+  std::size_t getMaxBytesPerBatch() const override {
+    return 128U;
+  }
+
+  std::size_t getMaxRetryCount() const override {
+    return 1U;
+  }
+
+  std::size_t getInitialRetryDelay() const override {
+    return 0U;
+  }
+
+  bool appendNewlineSeparators() const override {
+    return true;
+  }
+
+  std::size_t getFailedRecordCount(Outcome& outcome) const override {
+    return 0U;
+  }
+
+  Result getResult(Outcome& outcome) const override {
+    return outcome.GetResult().GetRecords();
+  }
+};
+
+TEST_F(AwsLoggerTests, test_send) {
+  auto log_forwarder = std::make_shared<DummyLogForwarder>();
+  log_forwarder->mustRun();
+
+  Dispatcher::addService(log_forwarder);
+
+  // The following 3 lines fit nicely inside a single batch
+  log_forwarder->logString("{ \"batch1\": \"1\" }");
+  log_forwarder->logString("{ \"batch1\": \"2\" }");
+  log_forwarder->logString("{ \"batch1\": \"3\" }");
+  std::this_thread::sleep_for(std::chrono::seconds(2));
```

Comment:
Can we remove these mandatory sleeps? Either replace with a function that polls extremely fast with a reasonable timeout, or refactor the logic to make it more testable.

Reply:
Done!